### PR TITLE
FAB-4903 initial commit of catalog management example

### PR DESCRIPTION
**Pull request by mcriscolo-cs**
ported the mx-disruptions example over to our examples repo.
See README.md in PR for more info on the example

### CODE REVIEW

**Review by laguirre-cs — COMMENTED** [review]
This example has a good amount of moving pieces. I think it would help to outline the main goal of the example is `Profile Creation with Catalog Management`. 

A diagram either explaining the code structure or components would be helpful. Here's a quick diagram I wrote up for this example (feel free to use it):

```mermaid
graph TD
    main_start[Start]-->startup[Startup - Create Cortex Session]
    startup-->Read
    subgraph Load JSON Config file - Catalog Resources
        %% Resources: Connections, Data Sources, Profile Schemas
        Read[Read Resources from Config]-->SafeDelete(Delete Resources, if they exist)
        SafeDelete-->Create[Create Resources, if they do not exist]
    end
    
    Create-- Resources -->Skip

    subgraph For Each Profile Schema in the Config File
        Skip{Skip Data Source Ingestion?}
        Skip-- Yes -->ConnectionProfileBuild[Create & Run BuildProfile Job<br>Using Connection as the Dataset]
        Skip-- No -->IngestPrimary[Create & Run IngestDataSource Job<br>Ingest Primary Data Source]

        %% Uncomment below to show a simpler view of the loop
        %%IngestPrimary-->IngestSecondary[Create & Run IngestDataSource Job<br>Ingest Secondary Data Sources]
        %%IngestSecondary-->ProfileBuild[Create & Run BuildProfile Job]
        
        IngestPrimary-->start
        subgraph Ingest Secondary Data Sources
            start-->condition
            condition{ProfileSchema has Joined DataSources?}-- yes -->effect[Create & Run IngestDataSource Job<br>Ingest Secondary Data Sources]
            effect-->start
        end
        condition -- no -->ProfileBuild[Create & Run BuildProfile Job]
    end
```

File: catalog-management/src/main/java/com/c12e/cortex/examples/catalog/ManageCatalog.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                BuildProfileJob buildProfileJob = cortexSession.job().buildProfile(project, profileSchemaName, cortexSession.getContext());
+                buildProfileJob.performFeatureCatalogCalculations = () -> false;
+                buildProfileJob.getDataset = (p, n) -> IngestDataSourceJob.DEFAULT_DATASOURCE_FORMATTER
+                        .apply(cortexSession.read().connection(p, cortexSession.catalog().getDataSource(p, n).getConnection().getName()).load());
+                buildProfileJob.run();
+            } else {
+                ProfileSchema profileSchema = cortexSession.catalog().getProfileSchema(project, profileSchemaName);
+
+                //build primary datasource
+                System.out.println("Ingesting Primary DataSource: " + profileSchema.getPrimarySource().getName());
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/java/com/c12e/cortex/examples/catalog/ManageCatalog.java:48]: done

File: catalog-management/src/main/java/com/c12e/cortex/examples/catalog/ManageCatalog.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    public void runApp(String project, CortexSession cortexSession, DocumentContext config) {
+        List<Map> profiles = config.read(APP_PATH + ".profiles");
+        Boolean skipDataSource = config.read("$.process.skipDataSource");
+            // Iterate over profiles in config
+        for (Map profile : profiles) {
+            // Build primary data source
+            String profileSchemaName = (String) profile.get("name");
+            if(skipDataSource) {
+                //build profile directly from connection, not supported for streaming connections
+                System.out.println("Building profile: " + profileSchemaName);
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/java/com/c12e/cortex/examples/catalog/ManageCatalog.java:38]: added to read me and code comment

File: catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+                    Arrays.asList(ctxt.readTreeAsValue(node.get("joins"), JoinSourceSelection[].class)),
+                    node.has("userId") ? node.get("userId").asText(null) : null,
+                    attributes,
+                    Arrays.asList(ctxt.readTreeAsValue(node.get("attributeTags"), AttributeTag[].class))
+            );
+        }
+    }
+
+    protected Boolean isProcessed(String checkpointDir) {
+        System.out.println("Checking for processed batch...");
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java:165]: removed

File: catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            //don't fail if not exists
+        }
+    }
+
+    public RailedCommand() {
+        SimpleModule module = new SimpleModule();
+        module.addDeserializer(ProfileSchema.class, new ProfileSchemaDeserializer());
+
+        com.jayway.jsonpath.Configuration.setDefaults(new com.jayway.jsonpath.Configuration.Defaults() {
+            private final JsonProvider jsonProvider = new JacksonJsonProvider(JsonMapper.builder()
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java:107]: added a comment and javadoc to RailedCommand and ManageCatalog

File: catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            Path file = new Path(checkpointDir + "/" + checkpointFileName);
+            return hdfs.exists(file);
+        } catch (IOException | URISyntaxException e) {
+            e.printStackTrace();
+        }
+        return false;
+    }
+
+    protected void writeCheckpoint(String checkpointDir) {
+        System.out.println("Writing checkpoint...");
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java:178]: removed

File: catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+                }
+                for (DataSource dataSource : dataSources) {
+                    safeDelete(() -> cortexSession.catalog().deleteDataSource(dataSource.getProject(), dataSource.getName()));
+                }
+                for (Connection connection : connections) {
+                    safeDelete(() -> cortexSession.catalog().deleteConnection(connection.getProject(), connection.getName()));
+                }
+            }
+        } catch (Exception e) {
+            //pass
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java:224]: removed unneeded try/catch

File: catalog-management/src/main/java/com/c12e/cortex/examples/catalog/ManageCatalog.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/java/com/c12e/cortex/examples/catalog/ManageCatalog.java:38]: done

File: catalog-management/README.md [hunk 7]
```diff
@@ -0,0 +1,304 @@
+# Managing Catalog Resources
+
+This example is a CLI application that uses a secondary configuration, the `app-config.json`, to define a number of
+catalog entities to be managed during execution. The connections/data sources/profile schema in the app-config.json
+are created wtih the attributes defined in the configuration and are then available to use through the profiles-sdk
+This builds off of the [Local Clients](../local-clients/README.md) example for its initial setup.
+
+(See [RailedCommand.java](./src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java) for config driven catalog management.)
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/README.md:8]: I don't think so in this instance, may be wrong @smichalski-cs

File: catalog-management/README.md [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+  }
+}
+```
+
+Notes on the above example:
+* The `--master` and `--deploy-mode` have been set to run the Spark job in the Cortex (Kubernetes) Cluster.
+* The Cortex API Client URL and Secret Client URL are referring to services in Kubernetes Cluster.
+* The Spark Driver and Spark Executors (`"spark.executor.instances"`) have a 2g and 4g of memory respectively. **Adjust the amount of resources used for your cluster/data.**
+* The Cortex [Backend Storage configuration](../docs/config.md#cortex-backend-storage) is configured by the default remote  storage client implementation.
+* The `--secret` is set in the `app_command`.
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/README.md:304]: removed

File: catalog-management/README.md [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+This builds off of the [Local Clients](../local-clients/README.md) example for its initial setup.
+
+(See [RailedCommand.java](./src/main/java/com/c12e/cortex/examples/catalog/RailedCommand.java) for config driven catalog management.)
+
+(See [ManageCatalog.java](./src/main/java/com/c12e/cortex/examples/catalog/ManageCatalog.java) for config driven profile build job.)
+
+## Prerequisites
+The app-config.json has a number of attributes which change how the example behaves.
+
+To skip the data source ingest and instead just build profile directly from the defined connection
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/README.md:15]: mmmmk, I don't know why it feels wrong but I will uppercase these to match the rest of the examples repo

File: catalog-management/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/README.md:8]: done

File: catalog-management/README.md [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+```json
+{
+  "process": {
+    "skipDataSource": true
+  }
+}
+```
+
+The list profile schemas to run the profile build job on
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/README.md:24]: done

File: catalog-management/README.md [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    "recreate": true
+  }
+}
+```
+
+To define the specification for any number of connections, datasources, and profile schemas
+
+```json
+{
+  "resources": {
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/README.md:52]: matches local catalog schema, done

File: catalog-management/README.md [hunk 12]
```diff
@@ -0,0 +1,304 @@
+# Managing Catalog Resources
+
+This example is a CLI application that uses a secondary configuration, the `app-config.json`, to define a number of
+catalog entities to be managed during execution. The connections/data sources/profile schema in the app-config.json
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/README.md:4]: done

File: catalog-management/src/main/resources/spark-conf/spark-conf.json [hunk 13]
```diff
@@ -0,0 +1,36 @@
+{
+  "pyspark": {
+    "pyspark_bin": "bin/spark-submit",
+    "app_command": [
+      "build-profile",
```
[the remark this answers is not in the document]
Comment (mcriscolo-cs) [reply catalog-management/src/main/resources/spark-conf/spark-conf.json]: done

**Review by laguirre-cs — APPROVED** [review]

**Review by mchutani-CS — APPROVED** [review]

**Pull request merged**